File: nodejs/base/api/lib/routing/defaults/default-route-configurator.ts
import { McmaResource } from "@mcma/core";
import { DefaultRouteCollection } from "./default-route-collection";
import { DefaultRouteCollectionBuilder } from "./default-route-collection-builder";
import { DefaultRouteBuilder } from "./default-route-builder";

export class DefaultRouteConfigurator<TResource extends McmaResource, TResult> {
    constructor(
        private builder: DefaultRouteCollectionBuilder<TResource>,
        private routes: DefaultRouteCollection<TResource>,
        private routeBuilder: DefaultRouteBuilder<TResult>
    ) { }

    configure(configureRoute: (defaultRouteBuilder: DefaultRouteBuilder<TResult>) => void): DefaultRouteCollectionBuilder<TResource> {
        if (!this.routes.included.includes(this.routeBuilder)) {
            this.routes.included.push(this.routeBuilder);
        }
        configureRoute(this.routeBuilder);
        return this.builder;
    }

    add(): this {
        if (!this.routes.included.includes(this.routeBuilder)) {
            this.routes.included.push(this.routeBuilder);
        }
        return this;
    }
    
    remove(): this {
        const index = this.routes.included.indexOf(this.routeBuilder);
        if (index >= 0) {
            this.routes.included.splice(index, 1);
        }
        return this;
    }
}

File: dotnet/base/Mcma.Api/Routing/Defaults/DefaultRouteConfigurator.cs

using System;
using Mcma.Core;

namespace Mcma.Api.Routing.Defaults
{
    public class DefaultRouteConfigurator<TResource, TResult> where TResource : McmaResource
    {
        internal DefaultRouteConfigurator(
            DefaultRouteCollectionBuilder<TResource> builder,
            DefaultRouteCollection<TResource> routes,
            DefaultRouteBuilder<TResult> routeBuilder)
        {
            Builder = builder;
            Routes = routes;
            RouteBuilder = routeBuilder;
        }

        private DefaultRouteCollectionBuilder<TResource> Builder { get; }

        private DefaultRouteCollection<TResource> Routes { get; }

        private DefaultRouteBuilder<TResult> RouteBuilder { get; }

        public DefaultRouteCollectionBuilder<TResource> Configure(Action<DefaultRouteBuilder<TResult>> configure)
        {
            if (!Routes.Included.Contains(RouteBuilder))
                Routes.Included.Add(RouteBuilder);
            configure(RouteBuilder);
            return Builder;
        }

        public DefaultRouteCollectionBuilder<TResource> Add()
        {
            if (!Routes.Included.Contains(RouteBuilder))
                Routes.Included.Add(RouteBuilder);
            return Builder;
        }

        public DefaultRouteCollectionBuilder<TResource> Remove()
        {
            if (Routes.Included.Contains(RouteBuilder))
                Routes.Included.Remove(RouteBuilder);
            return Builder;
        }
    }
}